pallene/coder.lua

```diff
@@ -1398,8 +1408,12 @@ gen_cmd["CallStatic"] = function(self, cmd, func)
     local top = self:stack_top_at(func, cmd)
 
     local parts = {}
-    table.insert(parts, self:call_pallene_function(dsts, #func.captured_vars >= 1,
-        cmd.f_id, top, xs))
+
+    -- CallStatic is used in pallene-pallene calls. Any calls to higher order functions or
+    -- closures is always done using CallDyn at the moment, meaning CallStatic will never call
+    -- a closure with upvalues.
+    assert(#self.module.functions[cmd.f_id].captured_vars == 0)
+    table.insert(parts, self:call_pallene_function(dsts, cmd.f_id, top, xs))
```

Comment:
I recommend using optional arguments sparingly. In a dynamic language, it's harder to tell if a missing argument is intentional or if it is an accident. In this case I would prefer passing an explicit `false` for the missing argument.

That said, as I said elsewhere, I think it would be even better if the `cclosure` argument were always a string. We want to allow CallStatic for functions that have captured variables.

Reply:
About the `CallStatic` being able to pass upvalues, we'd also have to be able to find a toplevel function's `upvalue` list, provided we know it's `id`. To that end, I have a small question:

Is it always guaranteed that  `function_n` is the `n`th value associated with the `globals` userdata?
If so, then what do you think about exposing the index `globals` and then getting the cclosure from that table?
I am also wondering the amount of overhead this might add to calling toplevel functions that capture globals as upvalues, since we're doing an extra push and pop operation.

```c
int type = lua_getiuservalue(L, globals, n);
assert(type == LUA_TFUNCTION);
// get the `upvalue` list from stack top, then pop from the stack?
lua_pop(L, 1)
```

Reply:
> Is it always guaranteed that function_n is the nth value associated with the globals userdata?

I think we can get this info from `upvalue_of_function`.

I don't remember if we are already creating a closure for every toplevel function. In the past we did not, but I don't remember if we recently changed this.

> I am also wondering the amount of overhead this might add to calling toplevel functions that capture globals as upvalues, since we're doing an extra push and pop operation.

We should't need to use a push and pop operation to get a value from the `G`. We can access the uservalue array directly.

Look at the places in the code that use with ir.Value.Function to see how we're currently dealing with this.

Reply:
> I don't remember if we are already creating a closure for every toplevel function. 

We don't create a `CClosure` object for toplevel functions declared like `local function foo`, but we do for `local foo: () -> () = ...`.  I realize now that it might be better to start by fixing this inconsistency first.

However, once that's done, we would still only be able to refer to local functions inside of `$init`.